### DISCUSSION TITLE
Naming of Duties, Coprocesses, RemoteTasks

### DISCUSSION THREAD

**Discussion by dthain** [cooperative-computing-lab/cctools#3158] [TaskVine Q&A] [answered]
Now that the duties are converging into a reliable feature, let's discuss names. I think it will help if the terms that we use match what is used in related systems. So, I am thinking something like the following:

- **Task** (a thing that runs somewhere, with data assets as inputs and outputs)
- Coprocess becomes **Service** (a thing that runs continuously and shouldn't ordinarily terminate)
- Duty becomes **ServiceTask** (a special case of a Task that describes a Service and starts it running)
- RemoteTask becomes **???** ServiceInvocation would be accurate, but quite a mouthful.  Other ideas?

@David-Simonetti-ND @btovar what do you think?

**Comment by btovar** [comment]
`C` -> `python`
`vine_task` -> `Task`
`vine_worker_function` -> `WorkerFn`, or `WorkerFunction`
`vine_task_function` -> `TaskFn` or `TaskFunction`

`RemoteTask` -> `ApplyFn`,  `CallFn`, `CallFunction`, `InvokeFn`

Being explicit about function makes clear that these are things you invoke with arguments.

Not a fan of Service and ServiceTask, as it sounds that ServiceTask is what you feed a Service.

**Comment by dthain** [comment]
Hmm, maybe FunctionDefTask and FunctionCallTask?

**Comment by dthain** [comment, marked as answer]
Decided on the whiteboard today:

A **Function** invokes code in a **Library**, which runs in a **Coprocess** at the **Worker**.

Both **Function** and **Library** are special cases of a **Task**.
